File: src/tools/impl/shell.py
import asyncio
from ..base import Tool, ToolResult, ToolDefinition, ToolParameter


class ShellTool:
    definition = ToolDefinition(
        name="shell",
        description="Execute shell commands",
        parameters={
            "command": ToolParameter(type="string", description="The shell command to execute", required=True)
        }
    )
    
    async def execute(self, params: dict) -> ToolResult:
        command = params["command"]
        try:
            proc = await asyncio.create_subprocess_shell(
                command,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )
            stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=30)
            output = stdout.decode() or stderr.decode()
            return ToolResult(
                success=proc.returncode == 0,
                output=output
            )
        except asyncio.TimeoutError:
            return ToolResult(success=False, output="", error="Command timed out")
        except Exception as e:
            return ToolResult(success=False, output="", error=str(e))

File: src/tools/impl/shell.ts
import { exec } from 'child_process';
import { promisify } from 'util';
import type { Tool, ToolResult } from '../base.js';

const execAsync = promisify(exec);

export const shellTool: Tool = {
  definition: {
    name: 'shell',
    description: 'Execute shell commands',
    parameters: {
      command: {
        type: 'string',
        description: 'The shell command to execute',
        required: true
      }
    }
  },
  
  async execute(params: Record<string, unknown>): Promise<ToolResult> {
    const command = params.command as string;
    try {
      const { stdout, stderr } = await execAsync(command, { timeout: 30000 });
      return {
        success: true,
        output: stdout || stderr
      };
    } catch (error: unknown) {
      const err = error as Error;
      return {
        success: false,
        output: '',
        error: err.message
      };
    }
  }
};
